### PR TITLE
Feature: Added new MASK-STRING-SKIP-ENDS transformation

### PR DESCRIPTION
**Pull request by nishant8887**
## Context

We use pipelinewise at our organization and need a transformation that masks string except the start and end.

### Changes

Added a new MASK-STRING-SKIP-ENDS transformation.

## Checklist
- [ ] I have considered the impact of this change and added a [Change Classification](
https://transferwise.atlassian.net/wiki/spaces/EKB/pages/1401189673/Change+Classifications+and+Expectations) Label (`change:standard`, `change:impactful`, `change:emergency`)
- [ ] Change meets or does not compromise the [Baseline Security Requirements](https://transferwise.atlassian.net/wiki/spaces/EKB/pages/434929973/Baseline+Security+Requirements)

### CODE REVIEW

**Comment by nishant8887** [comment]
Have created a separate version bump PR:https://github.com/transferwise/pipelinewise-transform-field/pull/28

**Comment by nishant8887** [comment]
@koszti Do you usually merge PRs from contributors outside of your organization?

File: tests/unit/test_transform.py [hunk 1]
```diff
@@ -59,6 +59,18 @@ def test_mask_hidden(self):
             'hidden'
         )
 
+    def test_mask_skip_ends(self):
```
Comment (Samira-El) [inline tests/unit/test_transform.py:62]: can you add couple tests for usecases where value to mask is not long enough?

**Review by Samira-El — COMMENTED** [review]
Hey, thank you for this nice little feature, looks like a great transformation to support.

I have left few comments if you don't mind addressing them.

Thx

**Comment by Samira-El** [comment]
> @koszti Do you usually merge PRs from contributors outside of your organization?

Yes, we do :)

File: transform_field/transform.py [hunk 2]
```diff
@@ -80,6 +80,12 @@ def do_transform(record, field, trans_type, when=None):
             elif trans_type == "MASK-HIDDEN":
                 return_value = 'hidden'
 
+            # Transforms string input to masked version skipping first and last n characters, e.g. MASK-SKIP-ENDS-3
+            elif 'MASK-SKIP-ENDS' in trans_type:
+                skip_ends_n = int(trans_type[-1])
+                return_value = value if len(value) < (2 * skip_ends_n) \
+                    else value[:skip_ends_n]+"*"*(len(value)-(2*skip_ends_n))+value[-skip_ends_n:]
```
[the remark this answers is not in the document]
Comment (nishant8887) [reply transform_field/transform.py:87]: f string... You mean str.format right?

File: transform_field/transform.py [hunk 3]
```diff
@@ -80,6 +80,12 @@ def do_transform(record, field, trans_type, when=None):
             elif trans_type == "MASK-HIDDEN":
                 return_value = 'hidden'
 
+            # Transforms string input to masked version skipping first and last n characters, e.g. MASK-SKIP-ENDS-3
+            elif 'MASK-SKIP-ENDS' in trans_type:
+                skip_ends_n = int(trans_type[-1])
+                return_value = value if len(value) < (2 * skip_ends_n) \
```
[the remark this answers is not in the document]
Comment (nishant8887) [reply transform_field/transform.py:86]: Yup... Seems a good idea.

File: tests/unit/test_transform.py (hunk 1 above)
Comment (nishant8887) [reply tests/unit/test_transform.py:62]: Will add those.

File: transform_field/transform.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nishant8887) [reply transform_field/transform.py:87]: Making it MASK-STRING-SKIP-ENDS. Any thoughts?

File: README.md [hunk 4]
```diff
@@ -95,6 +95,7 @@ You need to define which columns have to be transformed by which method and in w
 * **MASK-DATE**: Replaces the months and day parts of date columns to be always 1st of Jan
 * **MASK-NUMBER**: Transforms any numeric value to zero
 * **MASK-HIDDEN**: Transforms any string to 'hidden'
+* **MASK-SKIP-ENDS**: Transforms string input to masked version skipping first and last n characters, e.g. MASK-SKIP-ENDS-3
```
[the remark this answers is not in the document]
Comment (nishant8887) [reply README.md]: @Samira-El Making it MASK-STRING-SKIP-ENDS. Any thoughts?

File: transform_field/transform.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nishant8887) [reply transform_field/transform.py]: Done

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Samira-El) [reply README.md]: Actually now that I slept on it, I think I see why you named it `MASK-SKIP-ENDS`, I first had an issue with the "ENDS" part.

I think we can revert back to `MASK-SKIP-ENDS` 🙈

File: transform_field/transform.py [hunk 5]
```diff
@@ -80,6 +80,13 @@ def do_transform(record, field, trans_type, when=None):
             elif trans_type == "MASK-HIDDEN":
                 return_value = 'hidden'
 
+            # Transforms string input to masked version skipping first and last n characters
+            # e.g. MASK-STRING-SKIP-ENDS-3
+            elif 'MASK-STRING-SKIP-ENDS' in trans_type:
+                skip_ends_n = int(trans_type[-1])
+                return_value = '*'*len(value) if len(value) <= (2 * skip_ends_n) \
+                    else f'{value[:skip_ends_n]}{"*"*(len(value)-(2*skip_ends_n))}{value[-skip_ends_n:]}'
```
Comment (Samira-El) [inline transform_field/transform.py:88]: there are several calls to `len`, can you call it once and save it to a variable  
```suggestion
                value_len = len(value)
                return_value = '*'*value_len if value_len <= (2 * skip_ends_n) \
                    else f'{value[:skip_ends_n]}{"*"*(value_len-(2*skip_ends_n))}{value[-skip_ends_n:]}'
```

**Review by Samira-El — CHANGES_REQUESTED** [review]

File: transform_field/transform.py (hunk 5 above)
Comment (nishant8887) [reply transform_field/transform.py]: Done

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (nishant8887) [reply README.md]: MASK-STRING-SKIP-ENDS also works?

**Comment by nishant8887** [comment]
@Samira-El Please review https://github.com/transferwise/pipelinewise-transform-field/pull/29 as well when you plan to merge this PR. Just a version bump and change-log PR.

**Comment by Samira-El** [comment]
Can you add this new transformation to Readme [here](https://github.com/transferwise/pipelinewise-transform-field#transformation-types)?

**Comment by nishant8887** [comment]
> Can you add this new transformation to Readme [here](https://github.com/transferwise/pipelinewise-transform-field#transformation-types)?

Have already added it in the branch.

**Review by Samira-El — APPROVED** [review]

**Pull request merged**